driver-core/src/main/com/mongodb/operation/AsyncQueryBatchCursor.java

```diff
@@ -243,17 +284,24 @@ private void handleGetMoreQueryResult(final AsyncConnection connection, final Si
 
         @Override
         public void onResult(final BsonDocument result, final Throwable t) {
+            ServerCursor localCursor = getCursor();
+            AsyncConnectionSource localConnectionSource = getConnectionSource();
             if (t != null) {
-                Throwable translatedException = t instanceof MongoCommandException
-                                                ? translateCommandException((MongoCommandException) t, cursor)
+                Throwable translatedException = (t instanceof MongoCommandException && localCursor != null)
+                                                ? translateCommandException((MongoCommandException) t, localCursor)
                                                 : t;
                 connection.release();
                 close();
                 callback.onResult(null, translatedException);
             } else {
-                QueryResult<T> queryResult = getMoreCursorDocumentToQueryResult(result.getDocument("cursor"),
-                                                                                connectionSource.getServerDescription().getAddress());
-                handleGetMoreQueryResult(connection, callback, queryResult);
+
+                if (isClosed() || localConnectionSource == null) {
+                    callback.onResult(null, GET_MORE_AFTER_CLOSE_EXCEPTION);
+                } else {
+                    QueryResult<T> queryResult = getMoreCursorDocumentToQueryResult(result.getDocument("cursor"),
+                            localConnectionSource.getServerDescription().getAddress());
```

Comment:
The locking is not removing all the race conditions.   If close() is called at just the wrong time, localConnectionSource.release() could be called between the time we get our method-local reference to it and the time we try to use it, and we'd end up trying to use an instance with a zero reference count, which will thrown an exception.

What if instead of all the locking we just get the ServerAddress from:

     connection.getDescription().getServerAddress()

We know that's good because we've retained a reference to it and haven't yet released it.